### ISSUE TITLE
Question: Terragrunt init performance

### ISSUE THREAD

**Issue by abillingsley** [gruntwork-io/terragrunt#2986]
**Context**
I am looking for advice related to improving the time it takes to init a larger terragrunt project. 

The terragrunt code is organized in the following way:

```
├── live/
│   ├── dev
│   │   ├── global
│   │   │   └── someterragruntmodule
│   │   │      └── terragrunt.hcl
│   │   ├── us-east-1
│   │   │   └── someterragruntmodule1
│   │   │      └── terragrunt.hcl
│   │   └── us-west-2
│   │   │   └── someterragruntmodule2
│   │   │      └── terragrunt.hcl
│   ├── qa
│   ├── ppr
│   ├── trn
│   └── prod
└── src/
    └── modules
```

Our project has around 145 modules within the scope of a single environment (e.g dev).  When terragrunt analyzes the dependency graph it can group these 145 modules into 7 groups.  The vast majority of the terragrunt modules have terraform sources that are local to this repository (e.g `${get_repo_root()}/src/modules//somemodule`)

By in large the terraform modules (in `src/modules`) share some common providers notably
* hashicorp/aws 5.40.0
* DataDog/datadog 3.37.0

**Problem**
We use MSFT Azure DevOps (ADO) for CI/CD and are currently using MSFT hosted agents for our builds.  Some time ago we ran into an issue where the MSFT hosted agents were running out of disk space, primarily because .terragrunt-cache would copy distinct copies of the provider binaries for each of the terragrunt modules during init.  The AWS provider in particular is > 500mb on disk and quickly we encountered disk space issues resulting from many many copies of that large binary.  

To resolve this issue we overwrote the `TF_PLUGIN_CACHE_DIR` to reduce the number of provider binaries stored on disk.  Making this change required that init have a `--terragrunt-parallelism` equal to `1` to avoid concurrent writes to the provider cache.

Reducing terragrunt parallelism to `1` has significantly increased build times.  For example, the 145 modules for the dev environment will take around 40min to simply perform terragrunt init `terragrunt run-all init --terragrunt-working-dir ./live/dev --terragrunt-parallelism 1`

The slow init times aren't ideal for a number of reasons but we are quickly approaching the maximum time a MSFT ADO Hosted Agent job can run (60min by default)

**Things we've tried or considered**
* In an earlier version we had an ADO Job that would Init and Plan or Init and Apply but we have since created an ADO job that is simply responsible for init.  We then cache the results of the init process and restore that cache in subsequent jobs where we perform tasks like plan, validate, apply, etc.  We still incur the slow init times but this allowed us to extend the total build time beyond 60min because no one job is greater than 60 minutes (however the init job is now approaching that level)
* We are considering further partitioning the init process by region.  For example, we'd init `./live/dev/global`, `./live/dev/us-east-1` and `./live/dev/us-west-2` each in separate ADO jobs.  This will extend the time we have for perform init for each region to 60 minutes which will further increase our build times but also allow us to continue using MSFT hosted agents
* We are considering moving away from MSFT hosted agents so that we can
  * provision more disk (to enable parallelism during init)
  * remove the 60minute timeout

All of the things we've done or have considered doing are not solutions for the long term.  In all cases, we will eventually run out of disk, or run up against a hard timeout.  Not to mention the feedback/iteration time in the CI/CD process is painful when it takes 40min+ to know if an issue exists in the terrafrom plan etc.

Are there any other options available to us to further cache or tune the init process to gain performance and decrease build times?  

**Versions**
- Terragrunt version: 0.55.12
- Terraform version: 1.7.4

**Comment by lorengordon** [comment]
Related discussion here: https://github.com/gruntwork-io/terragrunt/issues/2920

**Comment by levkohimins** [comment]
Resolved in [v0.56.4](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.56.4) release. Make sure to read [Provider Caching](https://terragrunt.gruntwork.io/docs/features/provider-cache/).

**Comment by abillingsley** [comment]
@levkohimins I am attempting to adopt Provider Caching in our project

e.g
```
terragrunt run-all plan -lock=false --terragrunt-provider-cache --terragrunt-provider-cache-dir ./.terragrunt --terragrunt-working-dir ./live/dev
``` 

I have a module that uses a module from cloudposse (e.g)
```
module "naming" {
  source  = "cloudposse/label/null"
  version = "0.25.0"
  ...
 ```
 
 when running the command with the terragrund-provider-cache options I get an error like
 
 ```
 │ Error: Error accessing remote module registry
│ 
│   on ../XXXX/main.tf line 50:
│   50: module "naming" {
│ 
│ Failed to retrieve available versions for module "naming"
│ (../XXXX/main.tf:50) from registry.terraform.io: host
│ registry.terraform.io does not provide a modules service.
╵
```

any support is appreciated

**Comment by levkohimins** [comment]
> --terragrunt-provider-cache-dir ./.terragrunt

> │ Error: Failed to query available provider packages
│ 
│ Could not retrieve the list of available versions for provider
│ hashicorp/aws: cannot search ./.terragrunt: lstat ./.terragrunt: no such
│ file or directory
╵

@abillingsley, you got this error, because `./.terragrunt` doesn't exists.

**Comment by abillingsley** [comment]
except that it does exist.  Terragrunt creates it and it can be seen on the file system after the error occurs.  A subsequent run also with the directory already populated with cached providers also fails with the same issue

**Comment by levkohimins** [comment]
> when running the command with the terragrund-provider-cache options I get an error like
> 
> ```
> │ Error: Error accessing remote module registry
> │ 
> │   on ../XXXX/main.tf line 50:
> │   50: module "naming" {
> │ 
> │ Failed to retrieve available versions for module "naming"
> │ (../XXXX/main.tf:50) from registry.terraform.io: host
> │ registry.terraform.io does not provide a modules service.
> ```

@abillingsley, I can confirm there is such an issue. A fix release will be issues soon.

**Comment by levkohimins** [comment]
> except that it does exist. Terragrunt creates it and it can be seen on the file system after the error occurs. A subsequent run also with the directory already populated with cached providers also fails with the same issue

This is true, but the cache directory is created **only** when the `init` command needs to be run, this can happen either by explicitly specifying the `init` command or the `auto-init` feature. The error you showed tells us that the dir doesn't exist. Unfortunately, I was unable to reproduce this error.

**Comment by abillingsley** [comment]
> > except that it does exist. Terragrunt creates it and it can be seen on the file system after the error occurs. A subsequent run also with the directory already populated with cached providers also fails with the same issue
> 
> This is true, but the cache directory is created _only_ when the `init` command needs to be run, this can happen either by explicitly specifying the `init` command or the `auto-init` feature. The error you showed tells us that the dir doesn't exist. Unfortunately, I was unable to reproduce this error.

I can reproduce consistently when the custom terragrunt cache directory option is provided (`--terragrunt-provider-cache-dir`.  I can see if I can get more details for you but unfortunately, the logs provided are what I was given as output.  As of now, I am abandoning the custom cache directory option in favor of the default directory since that does not seem to exhibit the same issue.  My preference would be to use the custom cache directory for our CI/CD environment but it is not a deal breaker.

**Comment by levkohimins** [comment]
> > > except that it does exist. Terragrunt creates it and it can be seen on the file system after the error occurs. A subsequent run also with the directory already populated with cached providers also fails with the same issue
> > 
> > 
> > This is true, but the cache directory is created _only_ when the `init` command needs to be run, this can happen either by explicitly specifying the `init` command or the `auto-init` feature. The error you showed tells us that the dir doesn't exist. Unfortunately, I was unable to reproduce this error.
> 
> I can reproduce consistently when the custom terragrunt cache directory option is provided (`--terragrunt-provider-cache-dir`. I can see if I can get more details for you but unfortunately, the logs provided are what I was given as output. As of now, I am abandoning the custom cache directory option in favor of the default directory since that does not seem to exhibit the same issue. My preference would be to use the custom cache directory for our CI/CD environment but it is not a deal breaker.

I managed to reproduce this, we will fix it soon too.

**Comment by levkohimins** [comment]
In the meantime, you can specify the full path, everything should work `--terragrunt-provider-cache-dir /path/to/dir`

**Comment by levkohimins** [comment]
Resolved in [v0.57.1](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.57.1) release.
